src/main/kotlin/org/elm/ide/annotator/ElmUnresolvedReferenceAnnotator.kt

```diff
@@ -6,32 +6,69 @@ import com.intellij.psi.PsiElement
 import com.intellij.psi.PsiReference
 import org.elm.ide.intentions.ElmImportIntentionAction
 import org.elm.ide.intentions.ElmMakeDeclarationIntentionAction
+import org.elm.lang.core.psi.ElmFile
+import org.elm.lang.core.psi.ElmQID
 import org.elm.lang.core.psi.ancestors
 import org.elm.lang.core.psi.elements.*
+import org.elm.lang.core.resolve.reference.QualifiedConstructorReference
+import org.elm.lang.core.resolve.reference.QualifiedTypeReference
+import org.elm.lang.core.resolve.reference.QualifiedValueReference
 import org.elm.lang.core.resolve.reference.TypeVariableReference
 import org.elm.lang.core.resolve.scope.GlobalScope
+import org.elm.lang.core.resolve.scope.ImportScope
+import org.elm.lang.core.resolve.scope.ModuleScope
+
+
+/**
+ * A handler should return `true` if it handled the error and no further processing needs to be done.
+ */
+typealias BadRefHandler =
+        (ref: PsiReference, element: PsiElement, holder: AnnotationHolder) -> Boolean
+
 
 class ElmUnresolvedReferenceAnnotator : Annotator {
 
+    // A chain of handlers to be executed sequentially when a reference cannot be resolved.
+    // Order matters! Handlers earlier in the chain can short-circuit further evaluation.
+    private val handlers = listOf<BadRefHandler>(
+            ::handleTypeAnnotation,
+            ::handleSafeToIgnore,
+            ::handleModuleHiddenByAlias
+    )
+
     override fun annotate(element: PsiElement, holder: AnnotationHolder) {
         for (ref in element.references) {
             if (ref.resolve() == null) {
+                var handled = false
+                for (handler in handlers) {
```

Comment:
This could potentially be shortened to 
```kotlin
if (ref.resolve() == null && handlers.none { it(ref, element, holder) }) {
    holder.createErrorAnnotation(...)
}
```